### PR TITLE
mdbook documentation (#50)

### LINKED ISSUE #50: Use mdbook (or similar) for user documentation

### PR DESCRIPTION
**Pull request by dbanty**
This closes #50 by creating user documentation at https://triaxtec.github.io/dobby/ .

### Changes

1. Wrote a bunch of documentation in markdown under the /docs directory using mdBook
2. Pulled out a lot of user documentation from the code since that will live in the book now instead.
3. Changed any `pub` access to `pub(crate)` since `pub` was only used to generate documentation (and we can decide what to `pub` when we create an actual API).
4. Added an action to auto-publish the book to gh-pages on push to main.

### Things to Consider

1. Adding `prettier` or something to CI to keep our Markdown well formatted
2. Creating pull request templates which include a note about double checking that documentation is updated
3. Having some more complete examples of config files with some common or recommended workflows
4. Overall layout / structure. Most of the meat of the documentation is currently under the `config` heading which, while accurate, may not be the best UX.
5. Pulling some more basic info into the README. We don't want to duplicate knowledge but we can {{#include}} stuff from the README into the Book. Just not sure how much to put out there.

### CODE REVIEW

File: CODE_OF_CONDUCT.md [hunk 1]
```diff
@@ -14,22 +14,22 @@ appearance, race, religion, or sexual identity and orientation.
 Examples of behavior that contributes to creating a positive environment
```
Comment (sweetb) [inline CODE_OF_CONDUCT.md:14]: I like this file. Did you get this from somewhere?

**Review by sweetb — APPROVED** [review]
Looks good. I did not make sure no typos or spelling issues, and did not code review the rust specifically (though most changes appears to just be making a crate out of the top level packages). But I did actually read the docs on the endpoint, looks really good.

Some thoughts:

1. Is is worth "namespacing" commands (JiraIssueSelect, JiraIssueStatus vs. SelectJiraIssue, etc.)?
2. Could someone write their own dobby command just put it in a folder somewhere? The docs talk about writing something in python or rust and calling it from a Dobby command, but should we open it up so people can write the command themselves vs. calling it from shell?

File: CODE_OF_CONDUCT.md (hunk 1 above)
Comment (dbanty) [reply CODE_OF_CONDUCT.md:14]: Yes, this is the [Contributor Covenant Code of Conduct](https://www.contributor-covenant.org) which I've taken to adding to any open source projects. It's one of the options GitHub recommends in the Insights -> Community settings section for open source repos.

**Comment by dbanty** [comment]
> 1. Is is worth "namespacing" commands (JiraIssueSelect, JiraIssueStatus vs. SelectJiraIssue, etc.)?

Maybe? I like having them be somewhat english-readable, but it's probably at least worth organizing the docs in that way. `Command` is probably the only one that doesn't fall under another heading. Maybe "Other"? Then I think we have `Jira`, `GitHub`, `Git`, and something related to project management.

> 2. Could someone write their own dobby command just put it in a folder somewhere? The docs talk about writing something in python or rust and calling it from a Dobby command, but should we open it up so people can write the command themselves vs. calling it from shell?

There's no function to interpret Rust code at run time, you'd have to recompile Dobby, so you'd probably end up forking the project and writing your own command alongside the existing ones. Hopefully if it's worth integrating at the point it's also generic enough that someone would make a PR back to the project.

If you're using not-Rust to write something you really need to specify what's running it anyway, at which point I think the existing `Command` is best for flexibility.

**Comment by sweetb** [comment]
> There's no function to interpret Rust code at run time, you'd have to recompile Dobby, so you'd probably end up forking the project and writing your own command alongside the existing ones. Hopefully if it's worth integrating at the point it's also generic enough that someone would make a PR back to the project.

What about loading a shared object? That's how you'd typically do this. No need to do anything with this, just curious.

**Comment by dbanty** [comment]
> What about loading a shared object? That's how you'd typically do this. No need to do anything with this, just curious.

I don't think I understand what you mean. We can talk with mouth words if you have an idea for this.

**Pull request merged**